contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -1,436 +1,620 @@
-   extender performs binding instead of the scheduler. Extenders cannot be
-   invoked at other points, for example, they cannot be called before running
-   predicate functions.
-1. Every call to the extenders involves marshaling and unmarshalling JSON.
-   Calling a webhook (HTTP request) is also slower than calling native functions.
-1. It is hard to inform an extender that scheduler has aborted scheduling of
-   a Pod. For example, if an extender provisions a cluster resource and
-   scheduler contacts the extender and asks it to provision an instance of the
-   resource for the Pod being scheduled and then scheduler faces errors
-   scheduling the Pod and decides to abort the scheduling, it will be hard to
-   communicate the error with the extender and ask it to undo the provisioning
-   of the resource.
-1. Since current extenders run as a separate process, they cannot use
-   scheduler's cache. They must either build their own cache from the API
-   server or process only the information they receive from the default scheduler.
+Many features are being added to the Kubernetes Scheduler. They keep making the
+code larger and the logic more complex. A more complex scheduler is harder to
+maintain, its bugs are harder to find and fix, and those users running a custom
+scheduler have a hard time catching up and integrating new changes. The current
+Kubernetes scheduler provides [webhooks to extend](./scheduler_extender.md) its
+functionality. However, these are limited in a few ways:
+
+1.  The number of extension points are limited: "Filter" extenders are called
+    after default predicate functions. "Prioritize" extenders are called after
+    default priority functions. "Preempt" extenders are called after running
+    default preemption mechanism. "Bind" verb of the extenders are used to bind
+    a Pod. Only one of the extenders can be a binding extender, and that
+    extender performs binding instead of the scheduler. Extenders cannot be
+    invoked at other points, for example, they cannot be called before running
+    predicate functions.
+1.  Every call to the extenders involves marshaling and unmarshalling JSON.
+    Calling a webhook (HTTP request) is also slower than calling native
+    functions.
+1.  It is hard to inform an extender that scheduler has aborted scheduling of a
+    Pod. For example, if an extender provisions a cluster resource and scheduler
+    contacts the extender and asks it to provision an instance of the resource
+    for the pod being scheduled and then scheduler faces errors scheduling the
+    pod and decides to abort the scheduling, it will be hard to communicate the
+    error with the extender and ask it to undo the provisioning of the resource.
+1.  Since current extenders run as a separate process, they cannot use
+    scheduler's cache. They must either build their own cache from the API
+    server or process only the information they receive from the default
+    scheduler.
```

Comment:
Which of these deficiencies apply to custom schedulers ? People might have invested in those(extenders and custom schedulers) approaches and they might need to think about a migration path . Will those old mechanisms be deprecated eventually ?

Reply:
I wouldn't say these deficiencies precisely apply to custom schedulers. Rather, custom schedulers (those that are not extenders) are left without any of the features from the standard scheduler. They have to implement queuing, predicates, ranking, preemption, etc.